garbage collection vs usage tracking shutdown race condition

**Describe the bug**
I've seen this exception at least once in CI local tests. I think it comes from job completion garbage collection racing with DFK shutdown usage reporting, in examining the task dictionary.

```
parsl/tests/test_staging/test_elaborate_noop_file.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
parsl/dataflow/dflow.py:1051: in cleanup
    self.usage_tracker.send_message()
parsl/dataflow/usage_tracking/usage.py:224: in send_message
    message = self.construct_end_message()
parsl/dataflow/usage_tracking/usage.py:182: in construct_end_message
    app_fails = len([t for t in self.dfk.tasks if
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.0 = <dict_keyiterator object at 0x7f86107ca040>
>   app_fails = len([t for t in self.dfk.tasks if
                     self.dfk.tasks[t]['status'] in FINAL_FAILURE_STATES])
E   RuntimeError: dictionary changed size during iteration

```


**To Reproduce**
non-deterministic in CI

**Environment**
CI, near master 02d3b93ad4fd97b2411c000cf0b8820c76e5dfef

--- a/parsl/dataflow/usage_tracking/usage.py
+++ b/parsl/dataflow/usage_tracking/usage.py
@@ -10,7 +10,6 @@ import sys
 import platform
 import multiprocessing as mp
 
-from parsl.dataflow.states import FINAL_FAILURE_STATES
 from parsl.version import VERSION as PARSL_VERSION
 
 logger = logging.getLogger(__name__)
@@ -179,8 +178,7 @@ class UsageTracker (object):
 
         site_count = len([x for x in self.dfk.config.executors if x.managed])
 
-        app_fails = len([t for t in self.dfk.tasks if
-                         self.dfk.tasks[t]['status'] in FINAL_FAILURE_STATES])
+        app_fails = self.dfk.tasks_failed_count + self.tasks_dep_fail_count
 
         message = {'uuid': self.uuid,
                    'end': time.time(),

--- a/parsl/dataflow/usage_tracking/usage.py
+++ b/parsl/dataflow/usage_tracking/usage.py
@@ -178,7 +178,7 @@ class UsageTracker (object):
 
         site_count = len([x for x in self.dfk.config.executors if x.managed])
 
-        app_fails = self.dfk.tasks_failed_count + self.tasks_dep_fail_count
+        app_fails = self.dfk.tasks_failed_count + self.dfk.tasks_dep_fail_count
 
         message = {'uuid': self.uuid,
                    'end': time.time(),

